atd-toolbox/data_model/python/sql_templates/build_address_trigger.sql

```diff
@@ -0,0 +1,42 @@
+-- 
+-- function which creates a crash address based on address subparts
+--
+CREATE OR REPLACE FUNCTION build_crash_address(
+    block_num TEXT,
+    street_pfx TEXT,
+    street_name TEXT,
+    street_sfx TEXT
+)
+RETURNS TEXT AS $$
+DECLARE
+    address text := '';
+BEGIN
+    -- Concat each address part only if it's not null and not an empty string
+    IF block_num IS NOT NULL AND block_num <> '' THEN
+        address := block_num || ' ';
+    END IF;
+    
+    IF street_pfx IS NOT NULL AND street_pfx <> '' THEN
+        address := address || street_pfx || ' ';
+    END IF;
+    
+    IF street_name IS NOT NULL AND street_name <> '' THEN
+        address := address || street_name || ' ';
+    END IF;
+    
+    IF street_sfx IS NOT NULL AND street_sfx <> '' THEN
+        address := address || street_sfx || ' ';
+    END IF;
+    
+    -- Trim the final address to remove any trailing space
+    address := trim(address);
+
+    -- Return NULL if the trimmed address is empty
+     IF address = '' THEN
+        RETURN NULL;
+    ELSE
+        RETURN address;
+    END IF;
+
+END;
+$$ LANGUAGE plpgsql IMMUTABLE;
```

Comment:
i recently ran into this requirement for functions to be immutable when used in generated columns too! Cool to see how to handle this when defining a new function. I ran into it because `concat_ws` is not immutable in PG.

Reply:
this was a TIL for me too. i also learned that you can declare any postgres function to be `IMMUTABLE` and it will take your word for it 😬

Reply:
😬 indeed!